talk/basicconcepts/functions.tex

```diff
@@ -181,11 +181,11 @@
   \frametitlecpp[98]{Pass by value, reference or pointer}
   \begin{block}{Different ways to pass arguments to a function}
     \begin{itemize}
-    \item by default, arguments are passed by value (= copy) \\
-          good for small types, e.g.\ numbers
-    \item prefer references for mandatory parameters to avoid copies
-    \item use pointers for optional parameters to allow \mintinline{cpp}{nullptr}
-    \item use \mintinline{cpp}{const} for safety and readability whenever possible
+    \item By default, arguments are passed by value (= copy) \\
+          good for small types, e.g.\ anything smaller than a pointer
```

Comment:
Make it a bit larger please :) I think we argue later to also pass `std::span` and `std::string_view` by value, which are a pointer and a size_t. So I guess the guideline is: Pass by value what fits in a few registers.

I am also frequently passing `vecgeom::Vector<double>` by value now or `std::complex`. I think it follows a similar reasoning. But you can debate me! :)

Reply:
Yes, I tend to agree with you, but I wanted to give in a few words something beginners would easily understand, while having the [ABI](https://refspecs.linuxbase.org/elf/x86_64-abi-0.99.pdf#page=18) in mind. We can specify a size (i.e. up to 16 or 32 bytes), but I think that the size of a pointer is useful as advice because it's the size of a reference as well, and depending on what you are passing, it will be passed by reference anyway even when passed by value in code. I think I will also try to add a link to the ABI to the slide, for those who may want to check the details.

Reply:
Also keep in mind that there is no single ABI :) ARM and RISC-V are probably going to have their own rules for what is passed in registers and what on the stack. Windows also has it's own ABI and you can alter that with various calling conventions (e.g. `__vectorcall` and `__fastcall`). We should definitely not specify a size. I also tend to avoid pointing at a concrete ABI specification.

Btw, how is your ABI spec different than the Itanium C++ ABI here: https://github.com/itanium-cxx-abi/cxx-abi/? I thought the latter was the one that gcc/clang implement for Linux.

Reply:
I can remove the link, but amd64/x86_64 is pretty much the only thing people in the course would be using. The Itanium C++ ABI details only the C++ specific parts, it is built on top of the generic amd64/x86_64 ABI, which is not specific to C++ (see http://itanium-cxx-abi.github.io/cxx-abi/abi.html#docs). What do you recommend to say, then, if you don't want to give a specific size and don't want to use the pointer size (i.e. reference size) as a loose guideline?

Reply:
Remember that this is for the essentials course. The guideline needs to be very simple.
"numbers" is indeed a bit fuzzy, but something that is "a few bytes" or "about the size of a pointer" (== a few bytes) or "a primitive type" or a small struct should all be fine.

Reply:
If you want the "bigger" to be honoured, say "a few times the size of a pointer"

Reply:
I've dropped this commit from the changeset, let's address this separately.